### ISSUE TITLE
Use pretraining to avoid vanishing gradient in multi-layer networks

### ISSUE THREAD

**Issue by shamoons** [cazala/synaptic#23]
It seems that we should be able to pretrain networks of fewer layers before training a network with more layers. Or else we run into a vanishing gradient issue.

**Comment by cazala** [comment]
what do you mean by pre-training exactly?

**Comment by shamoons** [comment]
If you have a network with 1 input layer, 3 hidden layers, and one output layer, what if we first trained the network with 1 input layer, 1 hidden layer and the second hidden layer as the output.

Once it was trained sufficiently, we used the weights from Input -> Hidden1 as our starting weights and biases and then bolted on Hidden2. Trained that network with Hidden3 as the output.

Then repeat.

**Comment by cazala** [comment]
@shamoons Okay, I think I'm missing something here, what would be that target values to train the hidden layers (the ones you are using as output layer)? For example, let's say I have a network with 2 neurons in the input layer, 3 hidden layers of 10 neurons each, and 1 neuron in the output layer. I want to train this network an XOR, so my training set would be:

```
[ { input: [0,0], output: [0] },
  { input: [0,1], output: [1] },
  { input: [1,0], output: [1] },
  { input: [1,1], output: [0] } ]
```

So I take the first hidden layer as my only hidden layer, and the second one as the output layer. I can do an activation of that network, but I will get 10 outputs instead of one I don't see how can we compute some kind of error to make sure the network is actually making progress during the training. Also, when it's time to back-propagate, I have the final output that I want from the network (either 0 or 1) but I have to provide 10 target values to propagate from my output layer, so how do we figure out what are the correct targets?

**Comment by shamoons** [comment]
I guess I'm not sure of the solution myself. I'm addressing http://neuralnetworksanddeeplearning.com/chap5.html
